README.md
# ⏱ Unfog [![gh-actions](https://github.com/soywod/unfog/workflows/deployment/badge.svg)](https://github.com/soywod/unfog/actions?query=workflow%3Adeployment)

Minimalist CLI task & time manager, written in [Haskell](https://www.haskell.org).

![image](https://user-images.githubusercontent.com/10437171/89771094-1199da80-db00-11ea-8e65-12da9ec4161a.png)

## Table of contents

* [Motivation](#motivation)
* [Concept](#concept)
* [Installation](#installation)
* [Configuration](#configuration)
* [Usage](#usage)
* [Interfaces](#interfaces)
* [License](https://github.com/soywod/unfog/blob/master/LICENSE)
* [Changelog](https://github.com/soywod/unfog/blob/master/CHANGELOG.md)
* [Credits](#credits)

## Motivation

[Taskwarrior](https://taskwarrior.org) is a good and powerful CLI to track your
tasks, but it requires time to configure it. The amount of features is giant,
which can lead to confusion.  Plus, if you also want to track your time, you
need to install the [Timewarrior](https://taskwarrior.org/docs/timewarrior)
plugin, which makes the configuration step even heavier.

In the other hand, [Watson](https://github.com/TailorDev/Watson) tracks well
your time but can't tracks your tasks.

Unfog proposes an alternative solution that:

- is easy to install / configure / use
- can track your tasks AND your time

## Concept

A task is composed of a description and an optionnal due time. It can be
attached to a project. It can be started and stopped (to track time) and done
or deleted (to track projects).

Basic reports can be generated to have an overview of your tasks and your
projects. They allow you to determine how and where the time was spent, if
projects fits their initial estimations…

## Installation

```bash
curl -sSL https://raw.githubusercontent.com/soywod/unfog/master/install.sh | bash
```

*See [wiki section](https://github.com/soywod/unfog/wiki/Installation) for more
information.*

## Configuration

```toml
# ~/.config/unfog/config.toml

# Store file location.
# Default: $XDG_CONFIG_HOME/unfog/store
# Warning: only absolute paths are supported for now,
# see https://github.com/soywod/unfog/issues/45
store-path = "/abs/path/to/store"
```

*See [wiki section](https://github.com/soywod/unfog/wiki/Configuration) for
more information.*

## Usage

```
⏱ Unfog - Minimalist task & time manager

Usage: unfog COMMAND

Available options:
  -h,--help                Show this help text

Available commands:
  list                     Show current project tasks [l]
  info                     Show task details [i]
  worktime                 Show worktime report [wtime, w]
  status                   Show active task info [stat]
  add                      Add a new task [a]
  edit                     Edit an existing task [e]
  start                    Start a task [sta, s]
  stop                     Stop a task [sto, S]
  toggle                   Toggle a task [tog, t]
  done                     Mark as done a task [do, d]
  undone                   Unmark as done a task [undo, u]
  delete                   Delete a task [del, D]
  undelete                 Undelete a task [undel, U]
  context                  Change the current project [ctx, c]
  upgrade                  Upgrade the CLI
  version                  Show the version
  cache:clear              Clear the state cache
```

*See [wiki section](https://github.com/soywod/unfog/wiki/Usage) for more
information.*

## Interfaces

- [Unfog.vim](https://github.com/soywod/unfog.vim), a (neo)vim plugin

*See [wiki section](https://github.com/soywod/unfog/wiki/Interfaces) for more
information.*

## Credits

- [Kronos](https://github.com/soywod/kronos.vim), the unfog predecessor
- [Taskwarrior](https://taskwarrior.org), a task manager
- [Timewarrior](https://taskwarrior.org/docs/timewarrior), a time tracker plugin for Taskwarrior
- [Watson](https://github.com/TailorDev/Watson), a time tracker

*See [wiki section](https://github.com/soywod/unfog/wiki/Credits) for more
information.*


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.6] - 2023-03-07

### Fixed

- Incorrect table creation for `worktime --more` after introducing the PROJECT column [#58]

## [1.0.5] - 2022-12-13

- Also list PROJECT column with `worktime --more` [#56]

## [1.0.4] - 2021-07-08

### Added

- Add `--due-in` option to tasks list (reminders) [#41]

## [1.0.3] - 2020-12-04

### Changed

- Replace Travis CI by GitHub Actions

## [1.0.2] - 2020-12-04

### Added

- Config file `config.toml` + `store-path` option [#40]
- `cache:clear` procedure

### Changed

- Use GHC `v8.10.2`

### Fixed

- Task not found when using long ids containing dash [#43]
- Short ids for worktime --more query [#39]

## [1.0.1] - 2020-08-16

### Fixed

- List query as default command [#30]
- Completion [#37]
- Missing undelete command arg

## [1.0.0] - 2020-08-14

The args parser becomes more solid (based on
[optparse](https://github.com/pcapriotti/optparse-applicative#custom-parsing-and-error-handling)
lib). The due time is simplified and normalized (using ISO notation instead of
custom one), since it's not considered as a  main feature. Tags are replaced by
project. A project is just a unique tag (or a simple category). The aim of
those changes is to simplify the usage of Unfog, and to focus on the core
concept (create tasks, attach tasks to projects, start/stop tasks, check
reports).

### Added

- Save last state in a file [#25]

### Changed

- **[BREAKING]** Refactor id system [#33]
- **[BREAKING]** Refactor parser [#31]

### Fixed

- No parse due time [#34]

## [0.4.5] - 2020-07-27

### Changed

- Upgrade ghc `v8.8.4` to fix Windows build error

## [0.4.4] - 2020-07-27

### Added

- AUR package support [#26]
- Default command to list [#30]

### Changed

- Replace `Data.Duration` by own duration with tests [#24]
- Upgrade ghc `v8.8.3`

### Fixed

- Bad build state for update cmd [#27]

## [0.4.3] - 2020-03-11

### Added

- Init basic bash completion (commands, ids and tags) [#18]

## [0.4.2] - 2020-03-07

### Changed

- Remove secs and millis from queries [#23]
- All screenshots

### Fixed

- Bad wtime ordering [#22]
- Make install script more compatible by replacing `[[]]` with `[]`

## [0.4.1] - 2020-02-23

### Added

- `--more` option for worktime query

### Fixed

- `--json` output for worktime command

## [0.4.0] - 2020-01-11

### Added

- Upgrade command [#10]
- Add `status` command for UI integration
- Add `undone` command to unmark as done tasks [#20]

### Changed

- Shorter aliases
- Show command (UI) [#17]
- `worktime` command (group by ids) [#11]
- **[BREAKING]** Commands names + aliases

## [0.3.3] - 2020-01-01

### Added

- Colors [#4]

## [0.3.2] - 2019-12-30

### Added

- Refactor parsec exprs with `(<++)`
- Allow for batch processing of tasks [#13]

## [0.3.1] - 2019-12-28

### Fixed

- Active approx label (from `in ...` to `... ago`)

## [0.3.0] - 2019-12-28

### Added

- Due time [#9]

### Changed

- Make `+` optional in `context` and `worktime` commands [#16]

### Fixed

- Remove `print` [#15]

## [0.2.2] - 2019-12-22

### Added

- Worktime date range [#9]
- Due time [#9]

### Fixed

- Hyphen issue in desc [#12]

## [0.2.1] - 2019-12-08

### Added

- Add basic help command [#3]
- Add version command

### Changed

- Replace active ✔ by the elapsed time [#8]

### Fixed

- Delete command [#1]
- Worktime doesn't take in consideration the context [#2]

## [0.2.0] - 2019-12-01

### Added 

- Changelog
- New syntax to remove tags: `-tag` [#6]

### Changed

- **[BREAKING]** New parser introduced, based on parser combinators ([ReadP](https://hackage.haskell.org/package/base-4.12.0.0/docs/Text-ParserCombinators-ReadP.html)). Events have been impacted (simplified), so the previous store is not compatible anymore. You can upgrade it with this command: `sed -i -E 's/True |False |\+//g' ~/.config/unfog/store`

### Fixed 

- Create config dir if not exists [#5]

## [0.1.4] - 2019-11-26

### Added

- Installation script

### Changed

- Binaries compression (tar.gz)

## [0.1.3] - 2019-11-26

### Changed

- Build optimization

## [0.1.2] - 2019-11-26

### Added

- Travis CI
- Cross-compilation to Linux, OSX and Windows

## [0.1.1] - 2019-11-24

### Changed

- Update README and LICENSE

## [0.1.0] - 2019-11-23

First release :tada:

### Added

- Linux binaries

[unreleased]: https://github.com/soywod/unfog/compare/v1.0.5...HEAD
[1.0.5]: https://github.com/soywod/unfog/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/soywod/unfog/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/soywod/unfog/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/soywod/unfog/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/soywod/unfog/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/soywod/unfog/compare/v0.4.5...v1.0.0
[0.4.5]: https://github.com/soywod/unfog/compare/v0.4.4...v0.4.5
[0.4.4]: https://github.com/soywod/unfog/compare/v0.4.3...v0.4.4
[0.4.3]: https://github.com/soywod/unfog/compare/v0.4.2...v0.4.3
[0.4.2]: https://github.com/soywod/unfog/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/soywod/unfog/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/soywod/unfog/compare/v0.3.3...v0.4.0
[0.3.3]: https://github.com/soywod/unfog/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/soywod/unfog/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/soywod/unfog/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/soywod/unfog/compare/v0.2.2...v0.3.0
[0.2.2]: https://github.com/soywod/unfog/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/soywod/unfog/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/soywod/unfog/compare/v0.1.4...v0.2.0
[0.1.4]: https://github.com/soywod/unfog/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/soywod/unfog/compare/v0.1.2...v0.1.3

src/Table.hs
module Table where

import Data.List

type ColSize = Int

type Style = String

type Value = String

data Cell = Cell [Style] Value

renderCell :: ColSize -> Cell -> String
renderCell colSize (Cell styles val) = startStyle ++ val ++ padding ++ endStyle
  where
    startStyle = concat styles
    padding = replicate (colSize - length val + 1) ' '
    endStyle = "\x1b[0m"

renderCols :: [[Cell]] -> [[String]]
renderCols = map (intersperse sep) . transpose . map renderCols' . transpose
  where
    getCellVal (Cell _ val) = val
    renderCols' cells = map (renderCell colSize) cells
      where
        colSize = maximum (map (length . getCellVal) cells)

renderRows :: [[String]] -> [String]
renderRows = map ((++ "\n") . concat)

render :: [[Cell]] -> String
render = concat . renderRows . renderCols

-- Utils

sep :: String
sep = renderCell 0 (ext 8 . cell $ "|")

cell :: Value -> Cell
cell = Cell []

defineStyle :: Int -> Int -> Int -> Cell -> Cell
defineStyle color bright shade (Cell styles val) = Cell (styles ++ [style]) val
  where
    bright' = if bright > 0 then ";" ++ show bright else ""
    shade' = if shade > 0 then ";" ++ show shade else ""
    style = "\x1b[" ++ show color ++ bright' ++ shade' ++ "m"

reset = defineStyle 0 0 0

bold = defineStyle 1 0 0

underline = defineStyle 4 0 0

reversed = defineStyle 7 0 0

black = defineStyle 30 0 0

red = defineStyle 31 0 0

green = defineStyle 32 0 0

yellow = defineStyle 33 0 0

blue = defineStyle 34 0 0

magenta = defineStyle 35 0 0

cyan = defineStyle 36 0 0

white = defineStyle 37 0 0

brightBlack = defineStyle 30 1 0

brightRed = defineStyle 31 1 0

brightGreen = defineStyle 32 1 0

brightyellow = defineStyle 33 1 0

brightBlue = defineStyle 34 1 0

brightMagenta = defineStyle 35 1 0

brightCyan = defineStyle 36 1 0

brightWhite = defineStyle 37 1 0

bgBlack = defineStyle 40 0 0

bgRed = defineStyle 41 0 0

bgGreen = defineStyle 42 0 0

bgYellow = defineStyle 43 0 0

bgBlue = defineStyle 44 0 0

bgMagenta = defineStyle 45 0 0

bgCyan = defineStyle 46 0 0

bgWhite = defineStyle 47 0 0

bgBrightBlack = defineStyle 40 1 0

bgBrightRed = defineStyle 41 1 0

bgBrightGreen = defineStyle 42 1 0

bgBrightYellow = defineStyle 43 1 0

bgBrightBlue = defineStyle 44 1 0

bgBrightMagenta = defineStyle 45 1 0

bgBrightCyan = defineStyle 46 1 0

bgBrightWhite = defineStyle 47 1 0

ext = defineStyle 38 5
